// src/config/controladores/reviewController.js
const Review = require('../models/Review');
const Game = require('../models/Game'); // Necesario para verificar si el juego existe

// @route   POST api/reviews
// @desc    Crear una nueva reseña
// @access  Public (puedes añadir autenticación aquí)
exports.createReview = async (req, res) => {
  const { game, rating, comment } = req.body;

  try {
    // Verificar si el juego al que se asocia la reseña existe
    const existingGame = await Game.findById(game);
    if (!existingGame) {
      return res.status(404).json({ message: 'Juego no encontrado para esta reseña.' });
    }

    const newReview = new Review({
      game,
      rating,
      comment,
      // Si tienes un sistema de usuarios, podrías añadir user: req.user.id
    });

    const review = await newReview.save();
    res.status(201).json(review);
  } catch (err) {
    console.error(err.message);
    res.status(500).json({ message: 'Error del servidor al crear la reseña.' });
  }
};

// @route   GET api/reviews
// @desc    Obtener todas las reseñas
// @access  Public
exports.getReviews = async (req, res) => {
  try {
    // Populate para mostrar información del juego asociado
    const reviews = await Review.find().populate('game', 'title platform coverImage');
    res.json(reviews);
  } catch (err) {
    console.error(err.message);
    res.status(500).json({ message: 'Error del servidor al obtener las reseñas.' });
  }
};

// @route   GET api/reviews/:id
// @desc    Obtener una reseña por ID
// @access  Public
exports.getReviewById = async (req, res) => {
  try {
    const review = await Review.findById(req.params.id).populate('game', 'title platform coverImage');
    if (!review) {
      return res.status(404).json({ message: 'Reseña no encontrada.' });
    }
    res.json(review);
  } catch (err) {
    console.error(err.message);
    res.status(500).json({ message: 'Error del servidor al obtener la reseña.' });
  }
};

// @route   GET api/reviews/game/:gameId
// @desc    Obtener todas las reseñas de un juego específico
// @access  Public
exports.getReviewsByGameId = async (req, res) => {
    try {
      const reviews = await Review.find({ game: req.params.gameId }).populate('game', 'title platform coverImage');
      if (!reviews) {
        return res.status(404).json({ message: 'No se encontraron reseñas para este juego.' });
      }
      res.json(reviews);
    } catch (err) {
      console.error(err.message);
      res.status(500).json({ message: 'Error del servidor al obtener las reseñas del juego.' });
    }
  };


// @route   PUT api/reviews/:id
// @desc    Actualizar una reseña
// @access  Public (o privado si es por el propio usuario)
exports.updateReview = async (req, res) => {
  const { rating, comment } = req.body; // Solo permitimos actualizar rating y comment

  try {
    let review = await Review.findById(req.params.id);
    if (!review) {
      return res.status(404).json({ message: 'Reseña no encontrada.' });
    }

    // Aquí podrías añadir una verificación si la reseña pertenece al usuario que la está actualizando
    // if (review.user.toString() !== req.user.id) {
    //   return res.status(401).json({ message: 'No autorizado para actualizar esta reseña.' });
    // }

    review.rating = rating || review.rating;
    review.comment = comment || review.comment;

    await review.save();
    res.json(review);
  } catch (err) {
    console.error(err.message);
    res.status(500).json({ message: 'Error del servidor al actualizar la reseña.' });
  }
};

// @route   DELETE api/reviews/:id
// @desc    Eliminar una reseña
// @access  Public (o privado si es por el propio usuario)
exports.deleteReview = async (req, res) => {
  try {
    const review = await Review.findById(req.params.id);
    if (!review) {
      return res.status(404).json({ message: 'Reseña no encontrada.' });
    }

    // Aquí podrías añadir una verificación si la reseña pertenece al usuario que la está eliminando
    // if (review.user.toString() !== req.user.id) {
    //   return res.status(401).json({ message: 'No autorizado para eliminar esta reseña.' });
    // }

    await Review.findByIdAndDelete(req.params.id); // O review.remove() en versiones anteriores de Mongoose
    res.json({ message: 'Reseña eliminada con éxito.' });
  } catch (err) {
    console.error(err.message);
    res.status(500).json({ message: 'Error del servidor al eliminar la reseña.' });
  }
};
